### PR TITLE
DCS-1508 extend search engine to be able to search by last name only

### PR DESCRIPTION
**Pull request by michalnawrockidj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/SearchByLastName.kt [hunk 1]
```diff
@@ -2,11 +2,10 @@ package uk.gov.justice.digital.hmpps.welcometoprison.model.prison.prisonersearch
...
-  val firstName: String,
+data class SearchByLastName(
   val lastName: String,
-  val dateOfBirth: LocalDate,
+  val firstName: String? = null,
+  val dateOfBirth: LocalDate? = null,
   val pncNumber: String? = null,
-  val croNumber: String? = null,
   val nomsNumber: String? = null
 )
```
Comment (GurnankCheema) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/SearchByLastName.kt:11]: this makes lastName mandatory?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/validator/MatchPrisonersRequestValidator.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package uk.gov.justice.digital.hmpps.welcometoprison.validator
+
+import org.springframework.stereotype.Service
+import org.springframework.validation.Errors
+import org.springframework.validation.Validator
+import uk.gov.justice.digital.hmpps.welcometoprison.model.prison.prisonersearch.request.MatchPrisonersRequest
+
+@Service
+class MatchPrisonersRequestValidator : Validator {
+  override fun supports(clazz: Class<*>): Boolean {
```
Comment (GurnankCheema) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/validator/MatchPrisonersRequestValidator.kt:10]: clazz?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchApiClient.kt [hunk 3]
```diff
@@ -39,10 +39,10 @@ class PrisonerSearchApiClient(@Qualifier("prisonerSearchApiWebClient") private v
       ?: emptyList()
   }
 
-  fun matchPrisonerByNameAndDateOfBirth(searchByNameAndDateOfBirthOrPncNumber: SearchByNameAndDateOfBirth): List<Prisoner> {
+  fun matchPrisonerByName(searchByLastName: SearchByLastName): List<Prisoner> {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/PrisonerSearchApiClient.kt:42]: I think a name that makes it clear that we maybe matching by name and date of birth would be better?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/SearchByLastName.kt (hunk 1 above)
Comment (andrewrlee) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/SearchByLastName.kt:11]: probably worth removing noms and pnc number too

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/request/MatchPrisonersRequest.kt [hunk 4]
```diff
@@ -3,26 +3,22 @@ package uk.gov.justice.digital.hmpps.welcometoprison.model.prison.prisonersearch
 import com.fasterxml.jackson.annotation.JsonInclude
 import io.swagger.v3.oas.annotations.media.Schema
 import java.time.LocalDate
-import javax.validation.constraints.NotNull
 
 @JsonInclude(JsonInclude.Include.NON_NULL)
 @Schema(description = "Data for creating a search request to find mating prisoners")
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/request/MatchPrisonersRequest.kt:8]: typo in this

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/validator/MatchPrisonersRequestValidator.kt (hunk 2 above)
Comment (andrewrlee) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/validator/MatchPrisonersRequestValidator.kt:10]: that's a weird but common idiom in java - you can't uses the word `class` as its a keyword so people tend to use `clazz` instead!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/validator/MatchPrisonersRequestValidator.kt [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    val matchPrisonersRequest = target as MatchPrisonersRequest
+
+    if (matchPrisonersRequest.dateOfBirth == null &&
+      matchPrisonersRequest.prisonNumber == null &&
+      matchPrisonersRequest.pncNumber == null &&
+      matchPrisonersRequest.lastName == null
+    ) {
+      errors.rejectValue("lastName", "Last name is required")
+      errors.rejectValue("pncNumber", "PNC number is required")
+      errors.rejectValue("prisonNumber", "Last name is required")
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/validator/MatchPrisonersRequestValidator.kt:24]: The message `One of a,b,c must be provided` might be more helpful here?

Not sure you need to check for DoB also.. but worth holding off until we have chat

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/SearchByLastName.kt (hunk 1 above)
Comment (michalnawrockidj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/SearchByLastName.kt:11]: lets talk about this

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/request/MatchPrisonersRequest.kt (hunk 4 above)
Comment (michalnawrockidj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/prisonersearch/request/MatchPrisonersRequest.kt:8]: naughty naughty me ;-)

**Pull request closed without merging**